crates/sparrow-scheduler/src/pending.rs

```diff
@@ -0,0 +1,122 @@
+use std::sync::atomic::{AtomicBool, AtomicUsize, Ordering};
+use std::sync::Arc;
+
+use hashbrown::HashMap;
+use parking_lot::Mutex;
+
+use crate::Partition;
+
+/// Track the number of pending pipeline partitions.
+#[derive(Default)]
+pub(crate) struct PendingSet {
```

Comment:
Pulled this in from the prototype and added some logic to the scheduler as noted in worker. Just because the queue is empty doesn't mean we're "done". So instead, we need to track "which partitions are still running".

We *could* perhaps just do this for sources (make it part of the source tasks) but it seems nice to be able to say "which pipelines aren't completed" to diagnose things.

Reply:
Thoughts on complexity vs. value of tracking "what's blocking/outstanding"? It seems like we could just wait for the sources to be done and then wait for the queues to be empty.

Reply:
Main difficulty is figuring out how the pipelines communicate "done" back to the worker pool / manager.

This currently says: "each pipeline receives a `PendingPartition` instance which it can use to determine (a) if it's still pending and (b) report that it is complete. The worker then waits for all partitions to be done and can print which partitions (or the earliest pipelines) holding things up.

Alternatives:

1. Wait for all sources to be done then wait for the queue to be empty. This would require sources communicating to the worker pool that they are complete, but that shouldn't be too difficult since sources already exist outside the pool.
2. Have a special "PipelineEnd" that is a pipeline which when closed (or all partitions closed) signals the worker pool it's done. Have the `Write` actually close the pipeline end as it goes.